Add automated pre-/post-render helper functions

wlangera opened the pull request:
**Metadata**
- [x] update package version
- [x] update citation files
- [x] update NEWS
- [x] checklist checks
- [x] wait for v0.2.0 to be merged

**Functions**
`autoqmd_generate_children()`
- [x] script
- [x] input checks
- [x] documentation
- [x] examples
- [x] unit tests
- [x] improve code coverage

`autoqmd_insert_includes()` --> `autoqmd_insert_children()`
- [x] script
- [x] input checks
- [x] documentation
- [x] examples
- [x] unit tests
- [x] improve code coverage

`autoqmd_dependency_stamp()`
- [x] script
- [x] input checks
- [x] documentation
- [x] examples
- [x] unit tests

`autoqmd_prepare()`
- [x] script
- [x] input checks
- [x] documentation
- [x] examples
- [x] unit tests

`autoqmd_cleanup()`
- [x] script
- [x] input checks
- [x] documentation
- [x] examples
- [x] unit tests
- [x] improve code coverage

`autoqmd_finalise()`
- [x] script
- [x] input checks
- [x] documentation
- [x] examples
- [x] unit tests

**Other**
Tutorial
- [x] Review general text
- [x] Use iris dataset to make chapter per species example
- [x] Make freeze version of iris example

**Consistency**
- [x] Make sure examples and documentation follows good practises of tutorial
- [x] Make sure examples and documentation are consistent over autoqmd functions
       - `.qmd` files or QMD or Quarto
       - ...

**New tasks (see discussion below)**
- Update vignette
  - [x] remarks Els
  - [x] use better template file name example
  - [x] do not use `_qmd_` for freeze names
  - [x] add info about new argument dependency files
  - [x] explain use difference between quarto book and website?
- simplify `autoqmd_dependency_stamp`: users should give paths to dependencies themselves
  - [x] rewrite `autoqmd_dependency_stamp`
  - [x] add argument to `autoqmd_dependency_stamp`
  - [x] update unit tests of `autoqmd_insert_includes`
  - [x] add argument to `autoqmd_prepare`
- [x] Add warning message at the top of the automatically generated files (child files)
- [x] #12 (PR #13)

wlangera committed:
bump package version

wlangera committed:
update authors s.l.

wlangera committed:
update NEWS

wlangera committed:
add pkgdown website url

wlangera committed:
add pre render helper functions

wlangera committed:
add pkgdown website

wlangera committed:
clean documentation

wlangera committed:
add post-render functions

wlangera committed:
improve examples

wlangera committed:
use stopifnot where possible

wlangera committed:
same message as other functions

wlangera committed:
add checks

wlangera committed:
add delim argument

wlangera committed:
add internal validation checks

wlangera committed:
set up unit tests

wlangera committed:
set default for child dir

wlangera committed:
remove default

wlangera committed:
set up pkgdown article

wlangera committed:
improve documentation

wlangera committed:
add iris example for core workflow

wlangera committed:
finalise vignette article

wlangera committed:
fix check bug

wlangera committed:
create example qmds

wlangera committed:
create unit test

wlangera committed:
use copies of qmd file

wlangera committed:
add newline

wlangera committed:
make simple qmds on the fly

wlangera commented:
Issues:
- cyclomatic complexity of `autoqmd_dependency_stamp()` is too high
- create unit tests for `autoqmd_dependency_stamp()`
- in vignette refer to qmd files in testthat on github

wlangera committed:
fix linters

wlangera committed:
update buildignore

wlangera committed:
spelling check

wlangera committed:
review tutorial

wlangera committed:
refer correct function

wlangera committed:
add quiet argument

wlangera committed:
update documentation

wlangera committed:
check spelling

wlangera committed:
tidy description

wlangera committed:
remove motivation

wlangera committed:
remove non ascii code

wlangera committed:
correct url specification

wlangera committed:
spell check ignore docs

wlangera committed:
add vignetteBuilder

wlangera committed:
replace article with vignette

wlangera committed:
write checklist

wlangera committed:
Merge pull request #8 from inbo/checklist-vignette

Checklist vignette

wlangera committed:
use development version to check package

wlangera committed:
check with devel

wlangera committed:
test

wlangera committed:
try pak

wlangera commented:
@ThierryO can you review the `autoqmd` functionallity? We need to wait for v0.2.0 to be merged first.
I have provided code, documentation, unit tests, and a vignette.

I will also post 3 issues that I will not be working on in this PR:

> Issues:
> 
> * cyclomatic complexity of `autoqmd_dependency_stamp()` is too high
> * create unit tests for `autoqmd_dependency_stamp()`
> * in vignette refer to qmd files in testthat on github

Checklist checks are with the development version (0.5.2)

ThierryO commented on R/autoqmd_dependency_stamp.R:
```diff
@@ -0,0 +1,147 @@
+  env$here <- function(...) file.path(...) # minimal here()-like helper
+
+  # If rprojroot is available, provide the function directly
+  if (requireNamespace("rprojroot", quietly = TRUE)) {
```
We can require `rprojroot`. This removes the `if` statement and reduces the cyclomatic complexity

```suggestion
  stopifnot("`rprojroot`is not installed" = requireNamespace("rprojroot", quietly = TRUE))
```

ThierryO commented on R/autoqmd_dependency_stamp.R:
```diff
@@ -0,0 +1,147 @@
+  if (first_src > 1) {
+    assign_lines <- lines[seq_len(first_src - 1)]
+    assign_exprs <- try(parse(text = assign_lines), silent = TRUE)
+    if (!inherits(assign_exprs, "try-error")) {
```
Turn this in a clear error message instead of just a message.

ThierryO commented on R/autoqmd_dependency_stamp.R:
```diff
@@ -0,0 +1,147 @@
+    ln <- lines[i]
+    ex <- try(parse(text = ln)[[1]], silent = TRUE)
+    if (inherits(ex, "try-error")) {
+      if (!quiet) message("Could not parse line: ", ln)
```
copy the `n2kanalysis::display` function. The reduces the number of if statements.

ThierryO commented on R/autoqmd_dependency_stamp.R:
```diff
@@ -0,0 +1,147 @@
+    # The first arg of source(...) is ex[[2]] (could be a string,
+    # file.path(...), or variable)
+    path_expr <- try(ex[[2]], silent = TRUE)
+    if (inherits(path_expr, "try-error")) {
```
Would an error not be a better solution?

ThierryO reviewed (changes requested):
A quick review of the function with high cyclomatic complexity. Note that you probably need to split it over several smaller functions.

wlangera commented:
The function `autoqmd_dependency_stamp()` was created in the complex setting of the MAS report. The idea is to resimplify the function for this package. The use of rprojroot might not be so good with upcomming positron in mind.

steps:
- create unit tests
  - make different use cases, only basic ones (no rprojroot e.g.)
- simplify function while still passing tests
- pass checklist checks

wlangera committed:
update main package version

Merge branch 'main' of https://github.com/inbo/flandersqmd into 6-proposal-add-automated-quarto-prepost-render-helper-functions

# Conflicts:
#	.github/workflows/check_on_branch.yml
#	.github/workflows/check_on_main.yml
#	.zenodo.json
#	CITATION.cff
#	DESCRIPTION
#	NEWS.md
#	checklist.yml
#	inst/CITATION
#	inst/en_gb.dic

wlangera committed:
update checklist version

wlangera committed:
update checklist

wlangera committed:
update citation year

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,459 @@
+1. Generating one child `.qmd` file per entity (here: per species)
+2. Inserting `{{< include >}}` directives between the markers in the main chapter
+
+This can be done either explicitly or via a convenience wrapper.
```
It might be useful to give some advantages or disadvantages of each of the options to help users to make a choice.

wlangera replied:
Added in commit [6df3230](https://github.com/inbo/flandersqmd/commit/6df3230c964645d61b37d702d08657028cd91d5d)

ElsLommelen commented on vignettes/autoqmd.qmd:
````diff
@@ -0,0 +1,459 @@
+
+*Option 1: explicit generation and insertion*
+
+```r
````
For users that have limited experience with quarto and just use this manual to get it done, it might be useful to elaborate a bit (or provide a link) on this pre-render script: should it have a specific name?  Where should it be placed, in the same folder as the (q)md files or not? ...

wlangera replied:
Added in commit [f8d20fa](https://github.com/inbo/flandersqmd/commit/f8d20fa727f4d57234197f3fa04383417d05940a)

ElsLommelen reviewed (commented):
Just some minor suggestions while reading (and executing) the vignette.  I notice that the second suggestion/question will be answered in step 5, but it is much easier for a user to just follow the manual step by step without searching around.  ;-)
Up to now it is very clear in my opinion!

wlangera committed:
use underscore for template file name

wlangera committed:
do not use _qmd_ in case of freeze

wlangera committed:
simplify dependency time stamp function

wlangera committed:
create unit tests for dependency time stamps

wlangera committed:
correct imports assertthat

wlangera committed:
add template dependencies argument

wlangera committed:
more logic order of arguments

wlangera committed:
explain use of template deps argument

wlangera committed:
fix indentation

wlangera committed:
update book vs website explanation

wlangera committed:
give advantages and disadvantages for pre render options

wlangera committed:
explain pre and post render scripts

wlangera committed:
remove rprojroot dependency

wlangera committed:
Merge pull request #9 from inbo/simplify-timestamp-function

Simplify timestamp function

wlangera commented:
@ThierryO I have simplified the timestamp function. Users can now provide paths to their dependencies (e.g. R script or data file) if they want to freeze.

@ElsLommelen I have updated the vignette based on the discussion regarding dependencies (maandoverleg) and your comments

ElsLommelen commented on R/autoqmd_generate_children.R:
```diff
@@ -78,6 +78,8 @@ autoqmd_generate_children <- function(
   stopifnot("`delim` must be a character vector of length 2." =
               is.character(delim) && length(delim) == 2)
+  stopifnot("`quiet` must be a scalar logical vector." =
+              assertthat::is.flag(quiet))
```
Is there a reason why `assertthat::` is added?

wlangera replied:
To keep the validation checks short and simple. We already use assertthat in the package so it is not an extra dependency.
I just added this to check the user input, the argument type was not checked before.
Does this anwser what you mean?

``` r
# Logical vector
quiet <- c(TRUE, FALSE)

is.logical(quiet)
#> [1] TRUE
assertthat::is.flag(quiet)
#> [1] FALSE

# Scalar
quiet <- TRUE

is.logical(quiet)
#> [1] TRUE
assertthat::is.flag(quiet)
#> [1] TRUE
```

<sup>Created on 2026-01-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -24,7 +24,7 @@ At a high level, the workflow is:
 * **Clean up** autogenerated content and temporary files after rendering
 
-This approach is particularly useful for reports that must be reproducible, version-controlled, and regularly updated, such as monitoring reports or policy-oriented analyses. Conceptually, this approach treats repeated sections as temporary documents that are composed into the final report during rendering, rather than as loops or conditionals embedded inside a single `.qmd` file.
+This approach is particularly useful for reports and books that must be reproducible, version-controlled, and regularly updated, such as monitoring reports or policy-oriented analyses. Conceptually, this approach treats repeated sections as temporary documents that are composed into the final report during rendering, rather than as loops or conditionals embedded inside a single `.qmd` file. For Q ...
```
I'll do some more tests tomorrow or next week, but if I remember well: it can be usefut to produce the child documents (with freeze activated to make sure they always have the same name), but an additional functionality is needed to add these child documents on separate pages and the navigation tab.  (More precisely: the links to the child documents should be added to `_quarto.yml` in the prerender script).  So you might want to formulate more precisely what can be done and what can't?
Or you may want to leave it as is for now, I could add more info or the needed functionality after I get it working in my test case.

wlangera replied:
Thanks @ElsLommelen. Yes, I did not want to discuss this in too much detail as I think your type of website (currently) falls outside the scope of the `autoqmd` functions. I just want to emphasize how they work and the user should decide whether it is relevant or not.

If it is fine by you I would leave it as it is and we can expand this later (feel free to make an issue once you have enough tests). Unless you feel this text can be improved.

ElsLommelen reviewed (commented):
I looked at the recent changes without testing it, and I added some minor suggetions, up to you to decide if they are worth changing.
Thanks for addressing my previous remarks, from a user's point of view it is very clear now!
I may test it again in practice tomorrow or at the end of next week (but feel free to merge already if no problems are to be expected).

wlangera commented:
Version of this PR should be updated. v0.0.3 is on main

ElsLommelen commented:
Just some questions on this subject:

- When generating children with `autoqmd_generate_children()`, these children are added to the specified folder, but they are not supposed to be edited manually.  (When doing so, changes are overwritten anyway, but this might be confusing for users.)  Would it be an idea to automatically add a warning message on this (e.g. `"This file is generated automatically, do not make changes here, but in <xxx>"`, with `<xxx>` the name of the template) as a comment in these files while they are generated when running function `autoqmd_generate_children()`?

- What is the best practise: committing these files to GitHub or ignoring them as they can be generated automatically by building the book?

ThierryO commented:
> * When generating children with `autoqmd_generate_children()`, these children are added to the specified folder, but they are not supposed to be edited manually.  (When doing so, changes are overwritten anyway, but this might be confusing for users.)  Would it be an idea to automatically add a warning message on this (e.g. `"This file is generated automatically, do not make changes here, but in <xxx>"`, with `<xxx>` the name of the template) as a comment in these files while they are generated when running function `autoqmd_generate_children()`?

Adding such message at the top of the automatically generated files is a good idea. Look at what `roxygen2` with the `NAMESPACE` and `*.Rd` files

> * What is the best practise: committing these files to GitHub or ignoring them as they can be generated automatically by building the book?

IMHO you should ignore these file. I would add (or update) the `.gitignore` file in the same folder to ignore them automatically.

wlangera commented:
> > * When generating children with `autoqmd_generate_children()`, these children are added to the specified folder, but they are not supposed to be edited manually.  (When doing so, changes are overwritten anyway, but this might be confusing for users.)  Would it be an idea to automatically add a warning message on this (e.g. `"This file is generated automatically, do not make changes here, but in <xxx>"`, with `<xxx>` the name of the template) as a comment in these files while they are generated when running function `autoqmd_generate_children()`?
> 
> Adding such message at the top of the automatically generated files is a good idea. Look at what `roxygen2` with the `NAMESPACE` and `*.Rd` files

Good idea, I will add this.

> > * What is the best practise: committing these files to GitHub or ignoring them as they can be generated automatically by building the book?
> 
> IMHO you should ignore these file. I would add (or update) the `.gitignore` file in the same folder to ignore them automatically.

Yes they should be ignored. This is mentioned in the vignette under 'Customisation options and best practices' --> 'Cleanup of generated files'
Please let me know if this is not clear in the vignette

ElsLommelen commented:
> > IMHO you should ignore these file. I would add (or update) the `.gitignore` file in the same folder to ignore them automatically.
> 
> Yes they should be ignored. This is mentioned in the vignette under 'Customisation options and best practices' --> 'Cleanup of generated files' Please let me know if this is not clear in the vignette

Yes, you're right, it is in the vignette.  I forgot rereading this after adding `autoqmd_generate_children()`, thanks for mentioning!

ElsLommelen commented:
I have two more suggestions, up to you to decide on adding them (or not):
- when using child documents, it could be helpful to cache them, as it may be time consuming to build all of them each time during development while there are no changes.  For now I added this in the yml heading of my template qmd, but it may be more convenient to add it as an argument in `autoqmd_generate_children()`?  Otherwise it seems helpful to add it as a suggestion in the vignette.
- for use of child documents in a website, I developed code to replace a specific text in a `quarto.yml` template with titles for a TOC and references to the corresponding .qmd child documents (see code starting [here](https://github.com/inbo/forest-reserves/blob/eb98356fda5c88cf2e4c0023fe2338312074d845/source/website/pre_render.R#L22)).  It seems to work fine except for the fact that the preview is not running well after rendering due to [this reported issue](https://github.com/quarto-dev/quarto-cli/issues/4841).  Would it be an idea to add this as a separate function?  Please let me know if you want me to contribute in this.

wlangera commented:
Thanks for the nice ideas, @ElsLommelen!

For now, would you be willing to create an issue for each of your suggestions? For both, I’d like to think a bit more about the best way to implement them so everything stays consistent and well tested. You can reply to my comments below in the issues to provide any extra details.

> I have two more suggestions, up to you to decide on adding them (or not):
> 
> * when using child documents, it could be helpful to cache them, as it may be time consuming to build all of them each time during development while there are no changes.  For now I added this in the yml heading of my template qmd, but it may be more convenient to add it as an argument in `autoqmd_generate_children()`?  Otherwise it seems helpful to add it as a suggestion in the vignette.

Could you clarify what you mean by “now I added this in the yml heading”? A quick copy-paste or link would help. I’ll need to think about how this would work and how it differs from the existing freeze argument. If you have concrete ideas, feel free to share them in the issue.

> * for use of child documents in a website, I developed code to replace a specific text in a `quarto.yml` template with titles for a TOC and references to the corresponding .qmd child documents (see code starting [here](https://github.com/inbo/forest-reserves/blob/eb98356fda5c88cf2e4c0023fe2338312074d845/source/website/pre_render.R#L22)).  It seems to work fine except for the fact that the preview is not running well after rendering due to [this reported issue](https://github.com/quarto-dev/quarto-cli/issues/4841).  Would it be an idea to add this as a separate function?  Please let me know if you want me to contribute in this.

I agree it cwould be nice to have this as a separate function. I’d also like to think about how it would integrate with the other autoqmd functions in terms of naming and documentation. Also if you have a basic example to use in the unit tests, please let me know.

wlangera committed:
add warning to child files

wlangera committed:
update ror link

wlangera committed:
update package version

wlangera committed:
Merge branch 'main' into 6-proposal-add-automated-quarto-prepost-render-helper-functions

wlangera committed:
implement yaml method

wlangera committed:
update argument name for target file

wlangera committed:
add imports from tools

wlangera committed:
allow other arguments from autoqmd cleanup

wlangera committed:
add message for yaml files

wlangera committed:
update argument to target file

wlangera committed:
else position linter

wlangera committed:
add unit tests for yaml

wlangera committed:
make function name insert includes general

wlangera committed:
spelling check

wlangera committed:
add example for yaml

wlangera committed:
update documentation

wlangera committed:
split large function into helpers

wlangera committed:
add missing parenthesis

wlangera committed:
update vignette

wlangera committed:
test verbosity

wlangera committed:
test .yaml files

wlangera committed:
test YAML cleanup

wlangera committed:
Merge pull request #13 from inbo/12-add-childs-quarto-yml

Implement child documents for quarto website

wlangera committed:
correct message in vignette

wlangera commented:
@ElsLommelen I have implemented the use of `_quarto.yml`.
**Important:** I have changed the name `autoqmd_insert_includes()` --> `autoqmd_insert_children()` and the argument `qmd_file` to `target_file`.

The vignette is updated.

wlangera committed:
explain freeze argument in case of yaml workflow

ElsLommelen commented on R/autoqmd_finalise.R:
```diff
@@ -0,0 +1,77 @@
+  ...
+) {
+  # Validation
+  stopifnot("`target_files` must be a path to QMD files to update." =
+              all(sapply(target_files, assertthat::is.string)))
+  stopifnot("`target_files` must be a path to QMD files to update." =
+              all(sapply(target_files, file.exists)))
```
Bij het runnen van `autoqmd_finalise()` kreeg ik deze foutboodschap, vermoedelijk omdat het opgegeven bestand niet bestond, maar in deze context vond ik de foutboodschap zeer verwarrend: bij een path denk ik niet aan een bestand, enkel aan de bovenliggende mappen van het bestand, en in dit geval had ik enkel een bestand opgegeven.  Dus ben ik in de documentatie gedoken om te zien wat verwacht werd (ik vreesde dat behalve de naam ook de inhoud van het argument veranderd was), en daar vond ik dezelfde onduidelijke omschrijving van het path.  Anderzijds gaven voorbeelden de indruk dat ik een bestandsnaam inclusief extensie moest/kon opgeven (misschien kan dit mogeljik inclusief path?).
Enfin, de beschrijvingen van de foutmelding en functiedocumentatie dekken naar mijn aanvoelen niet de lading van wat verwacht (en getest) wordt, en idee is wel dat de gebruiker op basis daarvan genoeg info zou hebben (en niet in de code moet duiken om te zien wat het probleem zou kunnen zijn).

wlangera replied:
Ik ben niet zeker of ik het begrijp, dus even voor de duidelijkheid. Het gaat om dit argument:

```
target_files: Character vector. Paths to QMD files to update.
```

Een path naar een qmd file is toch telkens iets in de aard van `dir1/dir2/my_qmd.qmd`?
Dit moet een string zijn (zoals in de documentatie staat) en deze QMD file moet bestaan (in de documentatie staat "files to update", dus dit lijkt me wel logisch dat dit een bestaande file moet zijn).

Of bedoel je dat je per test graag een meer specifieke error message zou hebben?
Bijvoorbeeld:

- assertthat::is.string --> "`target_files` must be a character vector containing paths to QMD files to update."
-  file.exists  --> "At least one QMD file in `target_files` does not seem to exist."

wlangera replied:
Is dit al je volledige review of een opmerking tussen door? Want dan wacht ik met updates te doen tot je volledige review af is.
Merci

ElsLommelen replied:
> Een path naar een qmd file is toch telkens iets in de aard van `dir1/dir2/my_qmd.qmd`?
> Dit moet een string zijn (zoals in de documentatie staat) en deze QMD file moet bestaan (in de documentatie staat "files to update", dus dit lijkt me wel logisch dat dit een bestaande file moet zijn).

Voor mij is het path naar je qmd `dir1/dir2/`, de file `my_qmd.qmd` reken ik niet tot het path in strikte zin.  Ik zou dit dus eerder beschrijven als: 'qmd file including the path' als je effectief verwacht dat de gebruiker ook het bestand toevoegt.

> Of bedoel je dat je per test graag een meer specifieke error message zou hebben? Bijvoorbeeld:
> 
> * assertthat::is.string --> "`target_files` must be a character vector containing paths to QMD files to update."
> * file.exists  --> "At least one QMD file in `target_files` does not seem to exist."

Ja, het zou in mijn geval (file.exists) echt wel geholpen hebben, moest duidelijk aangegeven zijn dat de opgegeven file niet bestaat.  (In dit geval is het vrij aannemelijk dat de gebruiker een typfout gemaakt heeft ofzo, dan is het formaat in elk geval juist, dus dan helpt het niet echt om die algemene boodschap te herhalen.)

In het andere geval weet ik het niet zo goed: ik neem aan dat inderdaad wel kan helpen dat je aangeeft dat je een character vector verwacht.  Maar in die situatie heeft de gebruiker duidelijk niet door wat verwacht wordt, en ik vermoed dat de kans wel groot is dat hij/zij de documentatie niet gelezen heeft of verstrooid was, dus hier kan ik nog wel begrijpen dat je de omschrijving uit de documentatie herhaalt.   Anderzijds kan het wel goed zijn om andere bewoordingen te gebruiken dan in de documentatie of meer expliciet te zijn voor diegenen die de documentatie niet goed begrepen hebben.

(Dus eigenlijk was het de combinatie: enerzijds de error zelf die weinig specifiek was, anderzijds dat 'path' voor mij niet hetzelfde is als 'file' en ik dus verward was wat je als input verwachtte.)

> Is dit al je volledige review of een opmerking tussen door? Want dan wacht ik met updates te doen tot je volledige review af is.

Enkel een opmerking tussendoor waar ik gisteren toevallig bij uitkwam doen ik m'n code terug draaiend wou krijgen, ik zal vandaag de nieuwe werkwijze testen en dan een volledige review geven van mijn bevindingen (van het nieuwe deel).

wlangera replied:
Ok merci. Ik zal dat aanpassen en verbeteren.

Ik zal afwachten op je volledige review.

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+)
+
+# Insert entries into the Quarto configuration
+autoqmd_insert_includes(
```
```suggestion
autoqmd_insert_children(
```
I guess this must be changed?

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+        contents:
+        # AUTOGENERATED-START
+
+        # Navigation entries are autogenerated. Do not edit manually.
```
Wat is het nut van deze tekst?  Die verdwijnt blijkbaar bij het runnen van het pre-render script...

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+autoqmd_finalise(
+  target_files = "_quarto.yml",
+  message = message,
+  child_dirs_rm = "child_qmd",
```
Voor die `autoqmd_finalise()` ben ik niet zeker wat ik hiermee moet doen ingeval van freeze: als ik de functie toevoeg ben ik verplicht om argument `child_dirs_rm` aan te vullen (geen default waarde als ik dit niet wil), maar dan verlies ik uiteraard ook de rest van de functionaliteit.

Ook snap ik niet goed de keuze om bij `autoqmd_insert_children()` de zin `# Navigation entries are autogenerated. Do not edit manually.` weg te laten en te vervangen door tekst die niet aangepast mag worden, en nadien bij `autoqmd_finalise()` die automatisch gegenereerde tekst terug te verwijderen en te vervangen door die zin, die eigenlijk nietszeggend is als de gegenereerde tekst waar het over gaat, terug verwijderd wordt.  Zeker met de formulering van die zin zou het logischer zijn dat die weergegeven wordt op het moment dat de toegevoegde lijst er staat.

Zelf had ik m'n script zo gemaakt dat ik een template aanmaakte voor `_quarto.yml`, en het script die template kopieerde naar `_quarto.yml`, wat als voordeel heeft dat de originele file (=template) behouden blijft en de `_quarto.yml` read-only gemaakt kan worden, zodat er geen gevaar bestaat om dit te overschrijven.  Het feit dat de extra's toegevoegd worden binnen duidelijk gemarkeerde blokken in hetzelfde document, vind ik ook goed, maar ingeval ik freeze kies, zou het misschien logischer zijn dat behalve de files ook die lijst behouden blijft, met duidelijke aanduiding dat hier niks in aangepast mag worden.  Sowieso lijkt me dat ook handiger om problemen op te sporen, zeker voor diegenen die iets minder vertrouwd zijn met de machinerie (om te debuggen).

ElsLommelen replied:
Na nog eens herlezen van de sectie over freeze is me duidelijk geworden dat ik die cleanup beter niet doe, maar dan zou ik eigenlijk willen eindigen met een `_quarto.yml`-document waarin boven de auto-gegenereerde tekst die message staat dat daar niks in veranderd mag worden.  Dus die message die al vooraf in `_quarto.yml` toegevoegd was en die hier toegevoegd zou kunnen worden.

Waarschijnlijk is het handigst/properst/duidelijkst (meest begrijpelijk) als je die boodschap in het voorbeeld voor `_quarto.yml` in het vignet net boven de blok zetten, dan blijft die onveranderd en dan hoef je er codematig helemaal geen rekening mee te houden (ook niet hier bij de cleanup).
Alternatief om aan mijn vraag te voldoen, is het in `autoqmd_insert_children()` als argument toevoegen of de tekst in de blok behoudt van `_quarto.yml`, al lijkt dat codematig moeilijker en het is voor een gebruiker ook minder duidelijk als hij bij elke stap een nieuwe 'message' kan toevoegen en hij toch maar het eindresultaat met enkel de laatste boodschap gaat zien.  Dus ik zou zeggen: KISS: voeg in je voorbeeld een boodschap toe vlak boven het blok en dan lijkt het me niet meer nodig om bij verdere functies boodschappen in die blok toe te voegen.  En dan past de gebruiker de boodschap aan aan z'n situatie.

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+)
+
+# Insert includes into the chapter
+autoqmd_insert_includes(
```
```suggestion
autoqmd_insert_children(
```
Ik heb enkel de vignette online gelezen, dus misschien zoek je best eens in het hele package of de hernoemde functie en argument wel overal aangepast zijn?

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+
+my_species <- paste("Iris", levels(iris$Species))
+
+autoqmd_prepare(
+  species = my_species,
+  label = gsub("\\s", ".", tolower(my_species)),
+  template = "_species_template.qmd",
+  child_dir = "child_qmd",
+  target_file = "iris_chapter.qmd",
+  freeze = "label"
+)
```
Dat ziet er op zich wel een interessante functie uit, maar het grote nadeel t.o.v. `autoqmd_insert_children()` is voor mij dat het argument `child_labels` hier niet kan toegevoegd worden.  (Deze moeten net iets anders zijn dan de namen die ik uit de dataset haal.)  Is er een reden om die optie niet mee te geven in deze functie?

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+
+This timestamp records the modification time of the template used to generate the child files. It can be used to detect when a frozen child document may no longer be valid because its template has changed. 
+
+In addition to the template itself, users can specify external resources that influence the generated content, such as R scripts loaded with `source()` or data files read from disk. These can be declared via the `template_deps` argument in the `autoqmd_insert_includes()` or `autoqmd_prepare()` functions. When `template_deps` is supplied, modification times of the listed files are recorded alongsi ...
```
```suggestion
In addition to the template itself, users can specify external resources that influence the generated content, such as R scripts loaded with `source()` or data files read from disk. These can be declared via the `template_deps` argument in the `autoqmd_insert_children()` or `autoqmd_prepare()` functions. When `template_deps` is supplied, modification times of the listed files are recorded alongside the template timestamp. This makes all relevant inputs explicit and helps track when autogenerated content depends on files outside the template.
```
(And you may want to start a separate line for each new sentence, which makes it easier to track changes.)

ElsLommelen commented on R/autoqmd_insert_children.R:
```diff
@@ -0,0 +1,160 @@
+#' @param template Optional path to the QMD template used to generate
+#'   the child files. Its modification time will be
+#'   tracked.
+#' @param template_deps Optional character vector of file paths that are
+#'   explicit dependencies of the template (e.g. R scripts, data files).
+#'   Their modification times are included in the dependency stamp.
```
Mogelijk heb ik het niet goed begrepen, maar zou het niet logischer zijn om dit argument `template_deps` toe te voegen aan `autoqmd_generate_children()` in plaats van aan `autoqmd_insert_children()`?  Zoals ik het heb begrepen op basis van het vignet, bepaal je op die manier of (en welke?) children de volgende keer opnieuw gebouwd moeten worden, en het lijkt me wel logisch dat een gebruiker dit kan bepalen in `autoqmd_generate_children()`: iemand die enkel die documenten wil genereren, zonder toevoeging als lijstje in een document, gaat `autoqmd_insert_children()` niet nodig hebben, maar gaat wel graag die documenten enkel bouwen als ze nodig zijn.
En in je huidige configuratie worden die lijstjes van `autoqmd_insert_children()` toch telkens opnieuw gebouwd (en bij post-render verwijderd), dus ik snap niet goed welk voordeel dat argument `template_deps` hier kan hebben.

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+
+Freezing does not automatically disable cleanup. Whether child files are kept or removed after rendering depends on your post-render setup:
+
+* If `autoqmd_finalise()` removes the child directory, freezing has no effect across renders
```
Wat voor zin heeft het eigenlijk om freezing in te stellen als je een cleanup doet in de post-render stap?  Als hier geen voordeel bij is en die cleanup freezing ongedaan maakt zonder dat het nuttig is geweest, dan lijkt het erop dat de gebruiker het niet goed begrepen heeft of verstrooid is, dus dan zou ik een warning tijdens de cleanup wel op z'n plaats vinden.  Bij een warning loopt het script gewoon verder, dus de gebruiker heeft nog steeds de keuze om die warning te negeren, maar dan weet hij/zij tenminste dat die freeze niet gebruikt wordt.  (Je legt alles wel goed stap voor stap uit met voorbeelden, maar het blijft moeilijke materie en een beetje zoeken voor een gebruiker die dit de eerste keer toepast, dus warnings bij onlogisch gebruik kunnen enkel maar helpen om de gebruiker op de juiste weg te zetten.)

ElsLommelen commented on vignettes/autoqmd.qmd:
````diff
@@ -0,0 +1,662 @@
+When child documents are inserted into the main chapter, a dependency timestamp is added automatically:
+
+```text
+<!-- DEPENDENCY-MTIME: _species_template.qmd 2025-12-17 09:55:22 -->
````
Kan het zijn dat ik deze timestamp niet krijg bij een website?  (Ik vind hem nergens in mijn project.)  Heeft dit impact op de werking van dependencies?  (Ik zie dat de children niet opnieuw gebouwd worden, maar ik heb nog niet getest met aanpassingen in de template of toevoegen van `template_deps`.)

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+
+### Pre-render script
+
+The pre-render script generates child documents and inserts them into the YAML configuration. The `freeze` argument is used here to assign deterministic and meaningful filenames to the child documents; it does not invoke Quarto’s *freeze* functionality.
```
Euh, dat is wel verwarrend dat je hetzelfde argument `freeze` hier voor iets anders gebruikt.  Hoe kan ik dan wel 'freeze' instellen bij een website?  Of doet het beiden als je achteraf geen cleanup doet?  In dat geval ga je het hier wel heel anders moeten beschrijven/uitleggen: op z'n minst aangeven dat je freeze hier **bijkomend** gebruikt voor naamgeving, en dan uiteraard aangeven dat je een cleanup moet doen als je je files niet wil behouden (met link naar het juiste hoofdstuk waar hier meer uitleg over staat).  (En dat geeft ook een ander perspectief aan een eerdere opmerking.)

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+| -------- | -------------------------- | ----------------------------- | ---------------------------- |
+| Generate | Generated file location    | `autoqmd_generate_children()` | `child_dir`                  |
+| Generate | Verbosity                  | *all helpers*                 | `quiet`                      |
+| Insert   | Marker placement           | `autoqmd_insert_includes()`   | `start_marker`, `end_marker` |
+| Insert   | Page breaks                | `autoqmd_insert_includes()`   | `page_break`                 |
```
```suggestion
| Insert   | Marker placement           | `autoqmd_insert_children()`   | `start_marker`, `end_marker` |
| Insert   | Page breaks                | `autoqmd_insert_children()`   | `page_break`                 |
```

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+
+* Clearly named (e.g. `child_qmd/` or `generated/`)
+* Excluded from version control when appropriate
+* Fully removed during post-render cleanup
```
Dat laatste is afhankelijk van de situatie: bij grotere bestanden vraagt het telkens opnieuw genereren veel rekentijd, dus daar is de freeze-optie in combinatie met behouden van de files een betere optie.  Dus ofwel dat puntje wegdoen, ofwel beter kaderen met ook de optie van freeze?

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+  * Between sections in `.qmd` files
+  * Inside navigation structures such as `contents:` in `_quarto.yml`
+* Maintain correct indentation in YAML files to preserve valid syntax
+* Avoid nesting autogenerated blocks inside manually edited content
```
(laatste puntje): zou het dan niet beter zijn dat `autoqmd_insert_children()` een kopie maakt van de template die de gebruiker aanbiedt in plaats die file zelf aan te passen?  Dan kan de gegenereerde kopie read-only zijn en is er een kleiner risico dat er per ongeluk aanpassingen in die blocks gebeuren.  Sowieso kan het toch niet anders dat die autogenerated blocks in bv. een yaml binnenin manuele inhoud staat?  Dus in die zin snap ik niet goed hoe een gebruiker dat kan vermijden.

ElsLommelen commented on vignettes/autoqmd.qmd:
```diff
@@ -0,0 +1,662 @@
+* Maintain correct indentation in YAML files to preserve valid syntax
+* Avoid nesting autogenerated blocks inside manually edited content
+
+Clear and consistent markers make it obvious which parts of a document are safe to edit and which are managed automatically.
```
Afhankelijk van hoe je m'n andere commentaren aanpakt, hier evt. ook het belang van een duidelijke boodschap erboven vermelden.

ElsLommelen commented on vignettes/autoqmd.qmd:
````diff
@@ -0,0 +1,662 @@
+```text
+<!-- AUTOGENERATED-START -->
+
+> Autogenerated content below.
+> Do not edit manually.
````
Die autogenerated content staat er toch niet beneden op het moment dat iemand in die file werkt?

ElsLommelen reviewed (commented):
Ok, hier de beloofde uitgebreide review.
Algemeen gezien werkt die nieuwe functionaliteit goed: door mijn eerdere code te vervangen door je functie `autoqmd_insert_children()` krijg ik een `_quarto.yml` die er exact uitziet zoals het zou moeten.
Maar vanuit het oogpunt van een gebruiker heb ik hier en daar nog wel wat (detail)vragen bij de logica/begrijpbaarheid/gebruiksvriendelijkheid van de workflow, zeker vanuit mijn idee om met freeze te werken.  Dit gaat codematig over details en vooral over het communicatief wat beter en eenvoudiger te kunnen uitleggen, en de gebruiker de juiste knoppen aanbieden op het juiste moment.  Vaak kom ik een aantal keren op hetzelfde terug, soms met een iets andere invalshoek of nadat ik beter door had wat je bedoeling was, dus bekijk het best eerst allemaal en denk er even over na voor je in de code duikt.
Ook merkte ik hier en daar dat je je documentatie niet overal even goed aangepast hebt aan de nieuwe veranderingen, dus best even zoeken op de aangepaste namen en zorgen dat die overal aangepast zijn.

Misschien even kaderen: ik ben begonnen met in het vignet te lezen hoe ik de quarto-functionaliteit kon toevoegen, ik ben dan nog eens terug gaan zien hoe ik die freeze precies moest implementeren, en waar nodig ben ik verder naar antwoorden gaan zoeken in het eerste deel van het vignet of de functie-documentatie.  Dus ik heb niet alles in detail bekeken, ik heb het wel in de praktijk getest en het vignet vrij grondig bekeken met focus op het gebruik voor een website.  Bij deze denk ik dus wel dat ik de nieuwe functionaliteit vrij grondig bekeken heb, tenzij je nog andere dings toegevoegd zou hebben.  Maar anderzijds ben ik niet ingegaan op elk detail of argument, dus stel gerust andere argumenten ook nog in vraag die ik mogelijk niet bekeken heb.

wlangera committed:
update package metadata

wlangera committed:
Merge branch 'main' into 6-proposal-add-automated-quarto-prepost-render-helper-functions

wlangera committed:
update citation meta with checklist

ElsLommelen commented:
@wlangera,
While using function `autoqmd_generate_children()` together with `cache: true` in the template, I encountered an error that was very difficult to pinpoint.  The error was not dropped at the first run, but only in a new run after the template had changed.  The error message mentioned a non-existing problem (a conflict because it was not clear from which package function `filter()` should be run, while this problem was not present in the template ).
Removing folders `<childname>_cache` in  the `child_dir` and using `cache: false` turned out to be the solution, so most likely the error occurred because the old cache was not removed while a new `.qmd`-file was added.

Possible solutions:
- remove all subfolders starting with `<childname>` in function `autoqmd_generate_children()` for all `<childname>.qmd` that are replaced. (During rendering a folder `<childname>_files` is temporary added, and it stays sometimes there after a sudden error, so removing every file related to the replaced child has the advantage that nothing can cause problems, and it may be more robust if other related folders would be added in the future)
- the lazy solution: mention in the manual that cache cannot be used and drop an error if cache is set true.  This limits the options of the users, but as troubleshooting is hard here, I would not just drop a warning that can be lost somewhere while rendering.

wlangera commented:
Hi @ElsLommelen

I will try to replicate this.
Dit you set `cache: true` (1) for your whole Quarto project (in your `_quarto.yml`), (2) for a specific QMD chapter file, or (3) for (a) specific code chunck(s)?

Feel free to provide a link to your code.
